docs/the-new-architecture/pillars-cxxturbomodules.md

```diff
@@ -0,0 +1,698 @@
+
+```cpp
+struct CustomType {
+  std::string key;
+  bool enabled;
+  std::optional<int32_t> time;
+
+  bool operator==(const CustomType &other) const {
+    return key == other.key && enabled == other.enabled && time == other.time;
+  }
+};
+
+template <>
+struct Bridging<CustomType> {
+  static CustomType fromJs(
+      jsi::Runtime &rt,
+      const jsi::Object &value,
+      const std::shared_ptr<CallInvoker> &jsInvoker) {
+    return CustomType{
+        .key = bridging::fromJs<std::string>(
+            rt, value.getProperty(rt, "key"), jsInvoker),
+        .enabled = bridging::fromJs<bool>(
+            rt, value.getProperty(rt, "enabled"), jsInvoker),
+        .time = bridging::fromJs<std::optional<int32_t>>(
+            rt, value.getProperty(rt, "time"), jsInvoker)};
+  }
+
+  static jsi::Object toJs(jsi::Runtime &rt, const CustomType &value) {
+    auto result = facebook::jsi::Object(rt);
+    result.setProperty(rt, "key", bridging::toJs(rt, value.key));
+    result.setProperty(rt, "enabled", bridging::toJs(rt, value.enabled));
+    if (value.time) {
+      result.setProperty(rt, "time", bridging::toJs(rt, value.time.value()));
+    }
+    return result;
+  }
+};
+```
+
+Declare it in your `NativeSampleModule.h` file:
+
+```cpp
+  CustomType passCustomType(jsi::Runtime& rt, CustomType input);
+```
+
+Implement it in `NativeSampleModule.cpp`:
+
+```cpp
+CustomType NativeSampleModule::passCustomType(jsi::Runtime& rt, CustomType input) {
+  input.key = "1909";
+  input.enabled = !input.enabled;
+  input.time = 42;
+  return input;
+}
+```
+
+In your app you can call this new native function via:
+
+```js
+<Section title="Cxx TurboModule">
+  NativeSampleModule.passCustomType(...) ={' '}
+  {JSON.stringify(
+    NativeSampleModule.passCustomType({
+      key: '123',
+      enabled: true,
+      time: undefined
+    })
+  )}
+</Section>
+```
+
+which should return `{"key":"1909","enabled":false","time":42}`.
+
+This works - but is quite complex.
+
+### Struct generator
+
+[**Codegen**](pillars-codegen.md) for C++ Turbo Native Modules does support struct generators, so you can simplify the code above in `NativeSampleModule.h` to:
+
+```cpp
+using CustomType = NativeSampleModuleBaseCustomType<std::string, bool, std::optional<int32_t>>;
+template <>
+struct Bridging<CustomType>
+    : NativeSampleModuleBaseCustomTypeBridging<std::string, bool, std::optional<int32_t>> {};
+```
+
+With `using CustomType` you declare a name for your concrete struct.
+
+With `std::string, bool, std::optional<int32_t>` you define the property types of the struct members in the order they were defined in your JavaScript spec. By default, you can't `bridge` to other types for the first 2 arguments - but you can choose different JS `number` [representations in C++](https://github.com/facebook/react-native/blob/main/ReactCommon/react/bridging/Number.h).
```

Comment:
Is the order relevant? I guess so, in that case i will stress it with a bold syntax. I can see people opening issues because this doesn't work, where the problem is that they don't respected the ordering of the fields.

Reply:
> By default, you can't `bridge` to other types for the first 2 arguments

What does this mean?